### PR TITLE
support rotation of acrastructs encrypted without zones

### PR DESCRIPTION
**Pull request by Lagovas**
* published 1 function and 1 method from existing filesystem.KeyStore to reuse 
* extended rotator with zonemode parameter which turn on/off usage of key id as context for acrastructs

### CODE REVIEW

File: cmd/acra-rotate/rotator.go [hunk 1]
```diff
@@ -21,34 +21,34 @@ import (
...
 }
-func (rotator *keyRotator) getRotatedPublicKey(zoneID []byte) (*keys.PublicKey, error) {
-	keypair, ok := rotator.newKeypairs[string(zoneID)]
+func (rotator *keyRotator) getRotatedPublicKey(keyID []byte) (*keys.PublicKey, error) {
+	keypair, ok := rotator.newKeypairs[string(keyID)]
 	if ok {
 		return keypair.Public, nil
 	}
-	newKeypair, err := keys.New(keys.KEYTYPE_EC)
+	newKeypair, err := keys.New(keys.TypeEC)
```
Comment (ilammy) [inline cmd/acra-rotate/rotator.go]: ^_^ 👍

File: cmd/acra-rotate/keystore.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	return &rotateKeystore{store}, nil
+}
+
+// KeyStore interface used for acrastruct rotation
+type KeyStore interface {
+	SaveZoneKeypair([]byte, *keys.Keypair) error
+	SaveClientIDKeypair([]byte, *keys.Keypair) error
+	GetZonePrivateKey([]byte) (*keys.PrivateKey, error)
+	GetServerDecryptionPrivateKey([]byte) (*keys.PrivateKey, error)
+}
```
Comment (ilammy) [inline cmd/acra-rotate/keystore.go]: Is `rotateKeystore` intended to implement `KeyStore`? It seems to implement only `SaveClientIDKeypair()`.

File: tests/test.py [hunk 3]
```diff
@@ -3475,6 +3484,245 @@ def check_rotation(self, result, data_before_rotate, dry_run):
                 data_before_rotate[row['id']] = row['data']
 
 
+class TestAcraRotate(TestAcraRotateWithZone):
```
Comment (ilammy) [inline tests/test.py]: Your offering must be pleasant for the God of Quality. Thank you.

File: cmd/acra-rotate/rotator.go [hunk 4]
```diff
@@ -77,10 +77,55 @@ func (rotator *keyRotator) rotateAcrastruct(zoneID, acrastruct []byte) ([]byte,
 	return rotated, nil
 }
 
+func (rotator *keyRotator) rotateAcrastruct(id, acrastruct []byte) ([]byte, error) {
```
Comment (vixentael) [inline cmd/acra-rotate/rotator.go]: 👍

File: cmd/acra-rotate/rotator.go [hunk 5]
```diff
@@ -77,10 +77,55 @@ func (rotator *keyRotator) rotateAcrastruct(zoneID, acrastruct []byte) ([]byte,
...
+		logger.WithField("acrastruct", hex.EncodeToString(acrastruct)).WithError(err).Errorln("Can't get private key")
+		return nil, err
+	}
+	defer utils.FillSlice(0, privateKey.Value)
+	decrypted, err := base.DecryptAcrastruct(acrastruct, privateKey, nil)
+	if err != nil {
+		logger.WithField("acrastruct", hex.EncodeToString(acrastruct)).WithError(err).Errorln("Can't decrypt AcraStruct")
+		return nil, err
+	}
+	defer utils.FillSlice(0, decrypted)
```
Comment (vixentael) [inline cmd/acra-rotate/rotator.go]: 👍

File: cmd/acra-rotate/rotator.go [hunk 6]
```diff
@@ -77,10 +77,55 @@ func (rotator *keyRotator) rotateAcrastruct(zoneID, acrastruct []byte) ([]byte,
...
+func (rotator *keyRotator) rotateAcrastructWithClientID(clientID, acrastruct []byte) ([]byte, error) {
+	logger := log.WithFields(log.Fields{"ClientID": string(clientID)})
+	logger.Infof("Rotate AcraStruct")
+	// rotate
+	privateKey, err := rotator.keystore.GetServerDecryptionPrivateKey(clientID)
+	if err != nil {
+		logger.WithField("acrastruct", hex.EncodeToString(acrastruct)).WithError(err).Errorln("Can't get private key")
+		return nil, err
+	}
+	defer utils.FillSlice(0, privateKey.Value)
```
Comment (vixentael) [inline cmd/acra-rotate/rotator.go]: 👍

**Comment by vixentael** [comment]
'erge them all!

File: cmd/acra-rotate/keystore.go (hunk 2 above)
Comment (Lagovas) [reply cmd/acra-rotate/keystore.go]: it's inherit filesystem.KeyStore which implement all other methods and implement one more which needed by rotator

**Pull request merged**